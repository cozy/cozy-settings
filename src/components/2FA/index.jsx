import React, { Component } from 'react'
import { translate } from 'cozy-ui/react/I18n'

import Input from '../Input'
import PassphraseForm from '../PassphraseForm'
import Passphrase2FA from './Passphrase2FA'
import Activate2FA from './Activate2FA'
import Desactivate2FA from './Desactivate2FA'

const twoFaModalBanner = require('../../assets/images/double_authent_prez_banner.svg')
const twoFaModalProtect = require('../../assets/images/protect_data_point.svg')
const twoFaModalSecu = require('../../assets/images/niv_secu_point.svg')

class TwoFA extends Component {
  constructor (props) {
    super(props)
    this.state = {
      twoFAActivationModalIsOpen: false,
      twoFADesactivationModalIsOpen: false,
      twoFAPassphraseModalIsOpen: false,
      new2FAPassphrase: null,
      mailConfirmationCodeIsValid: false
    }
    // binding
    this.activate2FA = this.activate2FA.bind(this)
    this.desactivate2FA = this.desactivate2FA.bind(this)
    this.openTwoFAActivationModal = this.openTwoFAActivationModal.bind(this)
    this.closeTwoFAActivationModal = this.closeTwoFAActivationModal.bind(this)
    this.openTwoFADesactivationModal = this.openTwoFADesactivationModal.bind(this)
    this.closeTwoFADesactivationModal = this.closeTwoFADesactivationModal.bind(this)
    this.closeTwoFAPassphraseModal = this.closeTwoFAPassphraseModal.bind(this)
    this.onPassphrase2FAStep1 = this.onPassphrase2FAStep1.bind(this)
    this.onPassphrase2FASubmit = this.onPassphrase2FASubmit.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    this.props.twoFAField.value !== '' && nextProps.twoFAField.value && this.setState({mailConfirmationCodeIsValid: true})
  }

  activate2FA () {
    // TODO: Open the password modal
    this.props.activate2FA()
  }

  desactivate2FA () {
    this.props.desactivate2FA()
    // Reset all the info state
    this.props.updateInfo('two_fa', null)
    this.setState({
      mailConfirmationCodeIsValid: false
    })
    this.closeTwoFADesactivationModal()
  }

  openTwoFAActivationModal () {
    this.setState({twoFAActivationModalIsOpen: true})
  }
  closeTwoFAActivationModal () {
    this.props.cancel2FAActivation()
    this.setState({
      twoFAActivationModalIsOpen: false
    })
  }
  openTwoFADesactivationModal () {
    this.setState({twoFADesactivationModalIsOpen: true})
  }
  closeTwoFADesactivationModal () {
    this.setState({twoFADesactivationModalIsOpen: false})
  }
  closeTwoFAPassphraseModal () {
    this.setState((state, props) => ({
      twoFAPassphraseModalIsOpen: false,
      new2FAPassphrase: null
    }))
  }

  onPassphrase2FAStep1 (current, newVal) {
    this.setState((state, props) => ({
      twoFAPassphraseModalIsOpen: true,
      new2FAPassphrase: newVal
    }))
    this.props.onPassphrase2FAStep1(current)
  }

  onPassphrase2FASubmit (twoFactorCode) {
    const {
      onPassphrase2FAStep2,
      passphrase
    } = this.props
    const { twoFactorToken } = passphrase
    const { new2FAPassphrase } = this.state
    onPassphrase2FAStep2(new2FAPassphrase, twoFactorCode, twoFactorToken)
    .then(() => {
      this.closeTwoFAPassphraseModal()
    })
  }

  render () {
    const {
      t,
      twoFAField,
      passphrase,
      instance,
      checkTwoFactorCode,
      twoFactor
    } = this.props
    const {
      twoFAActivationModalIsOpen,
      twoFADesactivationModalIsOpen,
      twoFAPassphraseModalIsOpen,
      mailConfirmationCodeIsValid
    } = this.state
    const root = document.querySelector('[role=application]')
    const data = root.dataset
    const cozyDomain = data.cozyDomain
    return (
      <div>
        {twoFAField.value &&
          <PassphraseForm {...passphrase}
            onSubmit={this.onPassphrase2FAStep1}
          />
        }
        <Input
          name='two_fa'
          type='checkbox'
          title={t('ProfileView.twofa.title.activate')}
          label={t('ProfileView.twofa.label', {link: 'https://support.cozy.io/article/114-doubleauthentification'})}
          {...twoFAField}
          onChange={twoFAField.value
            ? this.openTwoFADesactivationModal
            : this.openTwoFAActivationModal
          }
        />
        {
          twoFAPassphraseModalIsOpen &&
          !passphrase.errors &&
          !passphrase.submitting &&
          <Passphrase2FA
            onPassphrase2FASubmit={this.onPassphrase2FASubmit}
            closeTwoFAPassphraseModal={this.closeTwoFAPassphraseModal}
            instance={instance}
            submitting={passphrase.submitting2FAStep2}
          />
        }
        {
          twoFAActivationModalIsOpen && <Activate2FA
            activate2FA={() => this.activate2FA()}
            checkTwoFactorCode={checkTwoFactorCode}
            mailConfirmationCodeIsValid={mailConfirmationCodeIsValid}
            closeTwoFAActivationModal={() => this.closeTwoFAActivationModal()}
            instance={instance}
            cozyDomain={cozyDomain}
            isTwoFactorEnabled={twoFAField.value}
            twoFactor={twoFactor}
            images={{
              'twoFaModalBanner': twoFaModalBanner,
              'twoFaModalSecu': twoFaModalSecu,
              'twoFaModalProtect': twoFaModalProtect
            }}
          />
        }
        {
          twoFADesactivationModalIsOpen && <Desactivate2FA
            desactivate2FA={() => this.desactivate2FA()}
            twoFactor={twoFactor}
            closeTwoFADesactivationModal={() => this.closeTwoFADesactivationModal()}
          />
        }
      </div>
    )
  }
}

export default translate()(TwoFA)
