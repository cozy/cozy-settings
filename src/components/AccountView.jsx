import styles from '../styles/accountView'

import React from 'react'
import ReactMarkdown from 'react-markdown'
import { translate } from 'cozy-ui/react/helpers/i18n'
import PassphraseForm from './PassphraseForm'
import Input from './Input'
import Select from './Select'

const LANG_OPTIONS = ['en', 'fr']

const AccountView = ({ t, fields, passphrase, isFetching, onFieldChange, onPassphraseSubmit }) => (
  <div className={styles['account-view']}>
    { isFetching && <p>Loading...</p> }
    <h2>{t('AccountView.title')}</h2>
    <Input name='email' type='email' {...fields.email} onChange={onFieldChange} />
    <Input name='public_name' type='text' {...fields.public_name} onChange={onFieldChange} />
    <Select name='locale' options={LANG_OPTIONS} {...fields.locale} onChange={onFieldChange} />
    <p className={styles['account-view-desc']}>
      <ReactMarkdown
        source={
          t(`AccountView.locale.contrib`)
        }
        renderers={{Link: props => <a href={props.href} target='_blank'>{props.children}</a>}}
      />
    </p>
    <PassphraseForm {...passphrase} onSubmit={onPassphraseSubmit} />
  </div>
)

export default translate()(AccountView)
