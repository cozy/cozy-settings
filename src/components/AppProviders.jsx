import React from 'react'
import { Provider, connect } from 'react-redux'
import { HashRouter } from 'react-router-dom'

import { BarProvider } from 'cozy-bar'
import { CozyProvider } from 'cozy-client'
import { DataProxyProvider } from 'cozy-dataproxy-lib'
import { WebviewIntentProvider } from 'cozy-intent'
import AlertProvider from 'cozy-ui/transpiled/react/providers/Alert'
import { BreakpointsProvider } from 'cozy-ui/transpiled/react/providers/Breakpoints'
import CozyTheme from 'cozy-ui/transpiled/react/providers/CozyTheme'
import I18n from 'cozy-ui/transpiled/react/providers/I18n'
import {
  StylesProvider,
  createGenerateClassName
} from 'cozy-ui/transpiled/react/styles'

import { PremiumProvider } from '@/components/Premium/PremiumProvider'
import { AvatarProvider } from '@/components/Profile/AvatarSection/AvatarContext'
import { FILES_DOCTYPE, CONTACTS_DOCTYPE, APPS_DOCTYPE } from '@/doctypes'

/*
With MUI V4, it is possible to generate deterministic class names.
In the case of multiple react roots, it is necessary to disable this
feature. Since we have the cozy-bar root, we need to disable the
feature.
https://material-ui.com/styles/api/#stylesprovider
*/
const generateClassName = createGenerateClassName({
  disableGlobal: true
})

const EnhancedI18n = connect(state => {
  const { lang } = state.ui
  return { lang }
})(I18n)

const AppProviders = ({ client, store, children }) => {
  return (
    <WebviewIntentProvider>
      <StylesProvider generateClassName={generateClassName}>
        <CozyProvider client={client}>
          <Provider store={store}>
            <EnhancedI18n
              dictRequire={lang => require(`@/locales/${lang}.json`)}
            >
              <BreakpointsProvider>
                <AlertProvider>
                  <CozyTheme className="u-w-100">
                    <DataProxyProvider
                      options={{
                        doctypes: [
                          FILES_DOCTYPE,
                          CONTACTS_DOCTYPE,
                          APPS_DOCTYPE
                        ]
                      }}
                    >
                      <BarProvider>
                        <PremiumProvider>
                          <AvatarProvider>
                            <HashRouter>{children}</HashRouter>
                          </AvatarProvider>
                        </PremiumProvider>
                      </BarProvider>
                    </DataProxyProvider>
                  </CozyTheme>
                </AlertProvider>
              </BreakpointsProvider>
            </EnhancedI18n>
          </Provider>
        </CozyProvider>
      </StylesProvider>
    </WebviewIntentProvider>
  )
}

export { AppProviders }
