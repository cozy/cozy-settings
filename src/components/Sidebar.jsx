import React from 'react'
import { translate } from 'cozy-ui/react/I18n'
import { NavLink as RouterLink } from 'react-router-dom'
import Nav, {
  NavLink,
  NavItem,
  NavIcon,
  NavText
} from 'cozy-ui/react/Nav'

import styles from '../styles/sidebar'

import boxIcon from '../assets/icons/icon-box.svg'
import globeIcon from '../assets/icons/icon-globe.svg'
import devicesIcon from '../assets/icons/icon-phone.svg'
import peopleIcon from '../assets/icons/icon-people.svg'

export const Sidebar = ({ t }) => (
  <aside className={styles['o-sidebar']}>
    <Nav>
      <NavItem>
        <RouterLink
          to='/profile'
          className={NavLink.className}
          activeClassName={NavLink.activeClassName}
        >
          <NavIcon icon={peopleIcon} />
          <NavText>{t('Nav.profile')}</NavText>
        </RouterLink>
      </NavItem>
      <NavItem>
        <RouterLink
          to='/connectedDevices'
          className={NavLink.className}
          activeClassName={NavLink.activeClassName}
        >
          <NavIcon icon={devicesIcon} />
          <NavText>{t('Nav.connected_devices')}</NavText>
        </RouterLink>
      </NavItem>
      <NavItem>
        <RouterLink
          to='/sessions'
          className={NavLink.className}
          activeClassName={NavLink.activeClassName}
        >
          <NavIcon icon={globeIcon} />
          <NavText>{t('Nav.sessions')}</NavText>
        </RouterLink>
      </NavItem>
      <NavItem>
        <RouterLink
          to='/storage'
          className={NavLink.className}
          activeClassName={NavLink.activeClassName}
        >
          <NavIcon icon={boxIcon} />
          <NavText>{t('Nav.storage')}</NavText>
        </RouterLink>
      </NavItem>
    </Nav>
  </aside>
)

export default translate()(Sidebar)
