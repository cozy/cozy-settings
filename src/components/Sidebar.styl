@require 'settings/breakpoints'

.Sidebar
    +gt-tablet()
        border-right 1px solid var(--dividerColor)
        min-width 320px
