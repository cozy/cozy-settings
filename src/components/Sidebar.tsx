import React from 'react'

import DevicesIcon from 'cozy-ui/transpiled/react/Icons/Devices'
import GlobeIcon from 'cozy-ui/transpiled/react/Icons/Globe'
import GraphCircle from 'cozy-ui/transpiled/react/Icons/GraphCircle'
import HandIcon from 'cozy-ui/transpiled/react/Icons/Hand'
import List from 'cozy-ui/transpiled/react/MuiCozyTheme/List'
import Logout from 'cozy-ui/transpiled/react/Icons/Logout'
import PeopleIcon from 'cozy-ui/transpiled/react/Icons/People'
import UnknowIcon from 'cozy-ui/transpiled/react/Icons/Unknow'
import flag from 'cozy-flags'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import { MenuItemAnchor } from 'components/menu/MenuItemAnchor'
import { MenuItemButton } from 'components/menu/MenuItemButton'
import { MenuItemNavLink } from 'components/menu/MenuItemNavLink'
import { MenuList } from 'components/menu/MenuList'
import { isFlagshipApp } from 'cozy-device-helper'
import { routes } from 'constants/routes'
import { useDiskPercentage } from 'hooks/useDiskPercentage'
import { useLogout } from 'hooks/useLogout'
import { useOffersLink } from 'hooks/useOffersLink'

export const Sidebar = (): JSX.Element => {
  const { t } = useI18n()
  const percent = useDiskPercentage()
  const logout = useLogout()
  const offersLink = useOffersLink()

  return (
    <nav role="navigation">
      <List>
        {isFlagshipApp() ||
          (flag('settings.flagship-mode') && (
            <MenuList title={t('Nav.header_flagship')}>
              <MenuItemNavLink
                to={routes.lockScreen}
                primary={t('Nav.primary_lock_screen')}
                icon={PeopleIcon} // @todo icon
              />
            </MenuList>
          ))}

        <MenuList title={t('Nav.header_general')}>
          <MenuItemNavLink
            to={routes.profile}
            primary={t('Nav.profile')}
            icon={PeopleIcon}
          />

          {flag('settings.enable_premium_links') && offersLink && (
            <MenuItemAnchor
              primary={t('Nav.primary_plan')}
              href={offersLink}
              target="_blank"
              icon={UnknowIcon}
            />
          )}

          <MenuItemNavLink
            to={routes.connectedDevices}
            primary={t('Nav.connected_devices')}
            icon={DevicesIcon}
          />

          <MenuItemNavLink
            to={routes.sessions}
            primary={t('Nav.sessions')}
            icon={GlobeIcon}
          />

          <MenuItemNavLink
            to={routes.storage}
            primary={t('Nav.storage')}
            secondary={
              percent &&
              t('Nav.secondary_used', {
                percent
              })
            }
            icon={GraphCircle}
          />

          {flag('settings.permissions-dashboard') && (
            <MenuItemNavLink
              to={routes.appList}
              primary={t('Nav.permissions')}
              icon={HandIcon}
            />
          )}
        </MenuList>

        <MenuList title={t('Nav.header_other')}>
          <MenuItemAnchor
            primary={t('Nav.primary_faq')}
            href={routes.external_faq}
            target="_blank"
            icon={UnknowIcon}
          />

          <MenuItemButton
            primary={t('Nav.primary_logout')}
            icon={Logout}
            onClick={(): void => void logout()}
          />
        </MenuList>
      </List>
    </nav>
  )
}

export default Sidebar
