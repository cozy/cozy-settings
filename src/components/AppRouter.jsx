import React from 'react'
import { Route, Navigate, Routes, Outlet } from 'react-router-dom'
import * as Sentry from '@sentry/react'

import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import flag from 'cozy-flags'

import Devices from 'containers/Devices'
import IntentRedirect from 'services/IntentRedirect'
import Passphrase from 'containers/Passphrase'
import PermissionsApplication from 'components/Permissions/AppPermissions/AppPermissions'
import PermissionsTab from 'components/Permissions/PermissionsTab'
import DataPermissions from 'components/Permissions/DataPermissions/DataPermissions'
import Support from 'components/Support/Support'
import PermissionDetails from 'components/Permissions/PermissionDetails/PermissionDetails'
import Profile from 'containers/Profile'
import Sessions from 'containers/Sessions'
import { routes } from 'constants/routes'
import ChangeEmail from 'components/Email/ChangeEmail'
import { DeleteAccount } from 'components/DeleteAccount'
import { Subscription } from 'components/Subscription/Subscription'
import { LockScreen } from 'components/pages/LockScreen'
import { Menu } from 'components/pages/Menu'
import { Storage } from 'components/Storage/Storage'

const SentryRoutes = Sentry.withSentryReactRouterV6Routing(Routes)

const OutletWrapper = ({ Component }) => (
  <>
    <Component />
    <Outlet />
  </>
)

const AppRouter = () => {
  const { isMobile, isTablet } = useBreakpoints()
  const isSmallView = isMobile || isTablet

  return (
    <SentryRoutes>
      {isSmallView && <Route path="/menu" element={<Menu />} />}
      <Route path="/redirect" element={<IntentRedirect />} />
      <Route path="/profile" element={<Profile />}>
        <Route path="/profile/email" element={<ChangeEmail />} />
        <Route path="/profile/delete" element={<DeleteAccount />} />
      </Route>
      <Route path="/profile/password" element={<Passphrase />} />
      <Route path="/connectedDevices/*" element={<Devices />} />
      <Route path="/connectedDevices/:deviceId" element={<Devices />} />
      <Route path="/sessions" element={<Sessions />} />
      <Route path="/storage" element={<Storage />} />
      <Route path={routes.lockScreen} element={<LockScreen />} />
      <Route path="/permissions/:page" element={<PermissionsTab />} />
      <Route
        path="/permissions/slug/:slug"
        element={<OutletWrapper Component={PermissionsApplication} />}
      >
        <Route path="details/:permissionType" element={<PermissionDetails />} />
      </Route>
      <Route path="/exports/:exportId" element={<Profile />} />
      <Route
        path="/permissions"
        element={<Navigate to="/permissions/slug" replace />}
      />
      <Route
        path="/permissions/data/:permissionType"
        element={<OutletWrapper Component={DataPermissions} />}
      >
        <Route path="details/:slug" element={<PermissionDetails />} />
      </Route>
      <Route path="/support" element={<Support />} />
      {flag('settings.subscription') && (
        <Route path="/subscription" element={<Subscription />} />
      )}
      <Route
        path="*"
        element={<Navigate to={isSmallView ? '/menu' : '/profile'} replace />}
      />
    </SentryRoutes>
  )
}

export { AppRouter }
