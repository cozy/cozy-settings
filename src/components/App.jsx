import styles from '../styles/app'

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { translate } from 'cozy-ui/react/I18n'
import classNames from 'classnames'
import { Route, Switch, Redirect, withRouter } from 'react-router-dom'

import Sidebar from './Sidebar'
import { Alerter } from 'cozy-ui/react/Alerter'
import Profile from '../containers/Profile'
import Devices from '../containers/Devices'
import Sessions from '../containers/Sessions'
import Storage from '../containers/Storage'

class App extends Component {
  render ({ children }) {
    return (
      <div className={classNames(styles['app-wrapper'], styles['o-layout--2panes'])}>
        <Alerter />
        <Sidebar />

        <main className={styles['app-content']}>
          <Switch>
            <Route path='/profile' component={Profile} />
            <Route path='/connectedDevices' component={Devices} />
            <Route path='/sessions' component={Sessions} />
            <Route path='/storage' component={Storage} />
            <Redirect exact from='/' to='/profile' />
            <Redirect from='*' to='/profile' />
          </Switch>
        </main>
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  alert: state.ui.alert
})

/*
withRouter is necessary here to deal with redux
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/guides/blocked-updates.md
*/
export default translate()(withRouter(connect(
  mapStateToProps
)(App)))
