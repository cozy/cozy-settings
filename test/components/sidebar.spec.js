'use strict'

/* eslint-env jest */

import React from 'react'
import { render } from '@testing-library/react'

import { Sidebar } from 'components/Sidebar'
import AppLike from '../AppLike'

describe('Sidebar component', () => {
  it('should be rendered correctly', () => {
    const root = render(
      <AppLike>
        <Sidebar />
      </AppLike>
    )
    root.getByText('Connected devices')
    root.getByText('Storage')
    root.getByText('Web connections')
  })
})
