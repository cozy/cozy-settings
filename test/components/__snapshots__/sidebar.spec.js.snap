// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component should be rendered correctly 1`] = `
<Sidebar
  className="o-sidebar"
>
  <Nav>
    <NavItem>
      <NavLink
        activeClassName="styles__is-active___2D0jN"
        ariaCurrent="true"
        className="styles__c-nav-link___3mK6W"
        to="/profile"
      >
        <NavIcon
          icon="test-file-stub"
        />
        <NavText>
          Profile
        </NavText>
      </NavLink>
    </NavItem>
    <NavItem>
      <NavLink
        activeClassName="styles__is-active___2D0jN"
        ariaCurrent="true"
        className="styles__c-nav-link___3mK6W"
        to="/connectedDevices"
      >
        <NavIcon
          icon="test-file-stub"
        />
        <NavText>
          Connected devices
        </NavText>
      </NavLink>
    </NavItem>
    <NavItem>
      <NavLink
        activeClassName="styles__is-active___2D0jN"
        ariaCurrent="true"
        className="styles__c-nav-link___3mK6W"
        to="/sessions"
      >
        <NavIcon
          icon="test-file-stub"
        />
        <NavText>
          Connections
        </NavText>
      </NavLink>
    </NavItem>
    <NavItem>
      <NavLink
        activeClassName="styles__is-active___2D0jN"
        ariaCurrent="true"
        className="styles__c-nav-link___3mK6W"
        to="/storage"
      >
        <NavIcon
          icon="test-file-stub"
        />
        <NavText>
          Storage
        </NavText>
      </NavLink>
    </NavItem>
  </Nav>
</Sidebar>
`;
